File: api/app/routers/matching.py
from typing import List

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from motor.motor_asyncio import AsyncIOMotorDatabase

from app.db import get_db
from app.models.schemas import ListingPublic
from app.services.matching import match_listings

router = APIRouter()


class MatchRequest(BaseModel):
    query: str
    zipCode: str | None = None


class MatchResponse(BaseModel):
    listings: List[ListingPublic]
    explanation: str


@router.post("/recommend", response_model=MatchResponse)
async def recommend(payload: MatchRequest, db: AsyncIOMotorDatabase = Depends(get_db)):
    # fetch listings
    rows = await db.listings.find({}).to_list(length=500)
    top, explanation = match_listings(rows, payload.query, payload.zipCode)
    return MatchResponse(listings=[ListingPublic(**l) for l in top], explanation=explanation)

File: web/src/api/matching.ts
import api from "./client";
import type { Listing } from "../types";

export async function recommend(input: { query: string; zipCode?: string }) {
  const { data } = await api.post<{
    listings: Listing[];
    explanation: string;
  }>("/matching/recommend", input);
  return data;
}

